static_frame/core/url.py

```diff
@@ -0,0 +1,94 @@
+import tempfile
+import typing as tp
+from io import BytesIO
+from io import StringIO
+from pathlib import Path
+from urllib import request
+from zipfile import ZipFile
+
+
+
+
+def url_adapter_file(
+        url: str,
+        encoding: tp.Optional[str] = 'utf-8',
+        in_memory: bool = True,
+        buffer_size: int = 8192,
+        ) -> tp.Union[Path, StringIO, BytesIO]:
+
+    with request.urlopen(url) as response:
+        if in_memory:
+            if encoding:
+                return StringIO(response.read().decode(encoding))
+            else:
+                return BytesIO(response.read())
+
+        # not in-memory, write a file
+        with tempfile.NamedTemporaryFile(mode='w' if encoding else 'wb',
+                suffix=None,
+                delete=False,
+                ) as f:
+            fp = Path(f.name)
+
+            if encoding:
+                extract = lambda: response.read(buffer_size).decode(encoding)
+            else:
+                extract = lambda: response.read(buffer_size)
+
+            while True:
+                b = extract()
+                if b:
+                    f.write(b)
+                else:
+                    break
+
+            return fp
+
+
+def url_adapter_zip(
+        url: str,
+        encoding: tp.Optional[str] = 'utf-8',
+        in_memory: bool = True,
+        buffer_size: int = 8192,
+        ) -> tp.Union[Path, StringIO, BytesIO]:
+
+    with request.urlopen(url) as response:
+        if in_memory:
+            archive = BytesIO(response.read())
+        else:
+            with tempfile.NamedTemporaryFile(mode='wb',
+                    suffix='zip',
+                    delete=False,
+                    ) as f:
+                archive = Path(f.name)
+                while True:
+                    b = response.read(buffer_size)
+                    if b:
+                        f.write(b)
+                    else:
+                        break
+    with ZipFile(archive) as zf:
+        names = zf.namelist()
+        if len(names) > 1:
+            raise RuntimeError(f'more than one file found in zip archive: {names}')
+        print(names)
+
+def URL(url: str,
+        encoding: tp.Optional[str] = 'utf-8',
+        in_memory: bool = True,
+        buffer_size: int = 8192,
+        ) -> tp.Union[Path, StringIO, BytesIO]:
```

Comment:
I'm not sure if this is a common use case, but it might be nice to be able to provide request headers here. I see [urlopen](https://docs.python.org/3/library/urllib.request.html#urllib.request.urlopen) allows the url to be a `urllib.request.Request` object which takes in headers. Alternatively maybe allowing the url to be a `Request` object could give that flexibility.

Reply:
Great point! I have adjusted the type hints and code to handle both an `str` or a `Request` object.